compiler/hash-ir/src/constant.rs

```diff
@@ -0,0 +1,537 @@
+            IntConstantValue::U32(val) => Scalar::from_uint(val, size),
+            IntConstantValue::U64(val) => Scalar::from_uint(val, size),
+            IntConstantValue::U128(val) => Scalar::from_uint(val, size),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Constructor)]
+pub struct ScalarInt {
+    /// The underlying scalar value.
+    data: Scalar,
+
+    /// Whether the integer is signed or not.
+    ty: IntTy,
+}
+
+impl fmt::Display for ScalarInt {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let size = self.data.size();
+        let bits = self.data.to_bits(size).unwrap();
+
+        if self.ty.is_signed() {
+            let value = size.sign_extend(bits) as i128;
+            write!(f, "{}_{}", value, self.ty.normalise(size))
+        } else {
+            write!(f, "{}_{}", bits, self.ty.normalise(size))
+        }
+    }
+}
+
+impl From<f32> for Scalar {
+    fn from(value: f32) -> Self {
+        Self {
+            value: value.to_bits() as u128,
+            size: NonZeroU8::new(std::mem::size_of::<f32>() as u8).unwrap(),
+        }
+    }
+}
+
+impl TryFrom<Scalar> for f32 {
+    type Error = Size;
+
+    fn try_from(value: Scalar) -> Result<Self, Self::Error> {
+        value
+            .to_bits(Size::from_bytes(std::mem::size_of::<f32>()))
+            .map(|u| f32::from_bits(u as u32))
+    }
+}
+
+impl From<f64> for Scalar {
+    fn from(value: f64) -> Self {
+        Self {
+            value: value.to_bits() as u128,
+            size: NonZeroU8::new(std::mem::size_of::<f64>() as u8).unwrap(),
+        }
+    }
+}
+
+impl TryFrom<Scalar> for f64 {
+    type Error = Size;
+
+    fn try_from(value: Scalar) -> Result<Self, Self::Error> {
+        value
+            .to_bits(Size::from_bytes(std::mem::size_of::<f64>()))
+            .map(|u| f64::from_bits(u as u64))
+    }
+}
+
+/// An [Alloc] represents a single allocation slot which contains the actual
+/// data, alignment, and mutability of the allocation.
+#[derive(Clone, Debug)]
+pub struct Alloc<Buf = Box<[u8]>> {
+    /// The buffer that is being used to store the value.
+    buf: Buf,
+
+    /// The alignment of the buffer.
+    align: Alignment,
+
+    /// The mutability of the allocation.
+    ///
+    /// ##Note: this is still not entirely figured out in the language. Perhaps, variables
+    /// that are 'static' can be mutable, and everything else is immutable.
+    mutability: Mutability,
+}
+
+/// Used for indexing into an [Alloc] by specifying the
+/// offset and size of the range.
+#[derive(Clone, Copy, Constructor)]
+pub struct AllocRange {
+    /// The offset into the allocation.
+    pub start: Size,
+
+    /// The size of the range.
+    pub size: Size,
+}
+
+impl AllocRange {
+    /// Get the end of the [AllocRange].
+    pub fn end(&self) -> Size {
+        self.start + self.size
+    }
+}
+
+pub trait AllocBuf:
+    Clone + fmt::Debug + Eq + PartialEq + Deref<Target = [u8]> + DerefMut<Target = [u8]>
```

Comment:
Why needing all of these bounds?

Reply:
- `Clone`, `Debug` are implied by the store
- `Eq` and `PartialEq` could be removed.
- `Deref` and `DerefMut` so the bytes can be directly accessed.